### PR TITLE
feat: add create command

### PR DESCRIPTION
**Pull request by Lanture1064**
#1

### CODE REVIEW

**Comment by bjwswang** [comment]
test robot

File: pkg/depository/create.go [hunk 1]
```diff
@@ -17,18 +17,177 @@ limitations under the License.
...
+				return err
+			}
+
+			// Marshal & Encode
+			msgJson, err := json.Marshal(msg)
+			if err != nil {
+				return err
+			}
+
+			//fmt.Printf("msgJson:\n%s\n---\n", msgJson)
```
Comment (Abirdcfly) [inline pkg/depository/create.go:145]: use V(5)?

**Comment by bjwswang** [comment]
@Lanture1064 test failed

**Comment by bjwswang** [comment]
> test robot

robot failed...

File: pkg/common/common.go [hunk 2]
```diff
@@ -21,3 +21,16 @@ const (
 	GetDepository    = "/basic/depositories/%s"
 	ListDepository   = "/basic/depositories"
 )
+
+type ValueDepository struct {
```
Comment (bjwswang) [inline pkg/common/common.go:25]: put component related moduels under its own package

File: pkg/depository/create.go [hunk 3]
```diff
@@ -17,18 +17,176 @@ limitations under the License.
...
+			// Get account address
+			var addr library.Address
+			err = addr.FromPublicKey(&privateKey.PublicKey)
+			if err != nil {
+				return err
+			}
+
+			klog.V(5).Infof("Getting nonce of account: %v\n", addr.String())
+
+			nonceURL := "http://localhost:9999/basic/currentNonce?account=" + addr.String()
```
Comment (bjwswang) [inline pkg/depository/create.go:101]: hard coded url ???

File: pkg/depository/create.go [hunk 4]
```diff
@@ -17,18 +17,176 @@ limitations under the License.
...
+
+			// Get nonce
+			resp, err := http.Get(nonceURL)
+			if err != nil {
```
Comment (bjwswang) [inline pkg/depository/create.go:105]: use common DoGetReq . pls work with @0xff-dev

File: pkg/depository/create.go [hunk 5]
```diff
@@ -17,18 +17,176 @@ limitations under the License.
...
+			rawVal, err := json.Marshal(valDep)
+			if err != nil {
+				return err
+			}
+
+			value := base64.StdEncoding.EncodeToString(rawVal)
+
+			// Obtain private & public key; generating key
+			// TODO: key auth
+			privateKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
```
Comment (bjwswang) [inline pkg/depository/create.go:87]: 1. we should save private key and public key to local file `~/.bestchains/{account_address}.json

2. add arguments for this command to use this key pair 
```
./bc-cli create depository --wallet ~/.bestchains --wallet-account 0xsklj34234fsf --wallet-account-generate
```
-  if account provided,we use `PutValue`  
- if --wallet-account-generate is specified , auto-generate account and use this account in `PutValue`
- if account not provided,we use `PutUntrustValue`

**Comment by bjwswang** [comment]
@Lanture1064 too many commits ...

**Review by bjwswang — APPROVED** [review]
/lgtm

File: pkg/depository/create.go [hunk 6]
```diff
@@ -17,18 +17,226 @@ limitations under the License.
...
+	cmd.Flags().StringP("platform", "p", "", "depot source platform")
+	cmd.Flags().StringP("wallet-account", "a", "", "wallet account address")
+	cmd.Flags().StringP("privateKey", "k", "", "private key of the given account address")
+	cmd.Flags().Bool("wallet-account-generate", false, "generates new account address")
+	cmd.Flags().Bool("untrusted", true, "put untrusted value")
+
 	return cmd
 }
+
+func generateValueDepotBase64(name string, contentType string, contentID string, platform string) string {
```
Comment (bjwswang) [inline pkg/depository/create.go:158]: add unit test for this kind of function

File: pkg/depository/create.go [hunk 7]
```diff
@@ -17,18 +17,226 @@ limitations under the License.
...
+	if err != nil {
+		return ""
+	}
+
+	value := base64.StdEncoding.EncodeToString(rawVal)
+
+	return value
+}
+
+func generateMessageBase64(nonce uint64, key ecdsa.PrivateKey, value []byte) string {
```
Comment (bjwswang) [inline pkg/depository/create.go:180]: add unit test for this kind of function

File: pkg/depository/create.go [hunk 8]
```diff
@@ -17,18 +17,226 @@ limitations under the License.
...
+	}
+
+	msgStr := base64.StdEncoding.EncodeToString(msgJson)
+	if err != nil {
+		return "Fatal: encode failed: " + err.Error()
+	}
+	return msgStr
+}
+
+func randAccountAndPrivateKey() (ecdsa.PrivateKey, library.Address) {
```
Comment (bjwswang) [inline pkg/depository/create.go:206]: add unit test for this kind of function

**Review by bjwswang — CHANGES_REQUESTED** [review]

File: pkg/depository/create.go [hunk 9]
```diff
@@ -17,18 +17,226 @@ limitations under the License.
...
+				non := getNonce(acc)
+				// generate msg
+				msgBase64 := generateMessageBase64(non, privKey, []byte(valueBase64))
+				// POST PutValue request
+				postHeader := map[string]string{
+					"message": msgBase64,
+					"value":   valueBase64,
+				}
+				resp, err := uhttp.Do(host, http.MethodPost, postHeader, nil)
+				if err != nil {
```
Comment (0xff-dev) [inline pkg/depository/create.go:106]: Has `API path` been added to host?

File: pkg/depository/create.go [hunk 10]
```diff
@@ -17,18 +17,226 @@ limitations under the License.
...
+				}
+				if !keyAuth(privKey) {
+					return fmt.Errorf("key Auth failed")
+				}
+				// get nonce
+				non := getNonce(acc)
+				// generate msg
+				msgBase64 := generateMessageBase64(non, privKey, []byte(valueBase64))
+				// POST PutValue request
+				postHeader := map[string]string{
```
Comment (0xff-dev) [inline pkg/depository/create.go:101]: The request parameters should be in the body, not in the request header.

File: pkg/depository/create.go [hunk 11]
```diff
@@ -17,18 +17,226 @@ limitations under the License.
...
+				privKey, addr := randAccountAndPrivateKey()
+				// Get Nonce
+				non := getNonce(addr.String())
+				// generate message
+				msgBase64 := generateMessageBase64(non, privKey, []byte(valueBase64))
+				// POST PutValue request
+				postValue := url.Values{}
+				postValue.Add("message", msgBase64)
+				postValue.Add("value", valueBase64)
+				host = fmt.Sprintf("%s?%s", common.CreateDepository, postValue.Encode())
```
Comment (0xff-dev) [inline pkg/depository/create.go:132]: I think it should be.
```go
host = fmt.Sprintf("%s%s?%s", host, common.CreateDepository, postValue.Encode())
```

File: pkg/depository/create.go [hunk 12]
```diff
@@ -17,18 +17,226 @@ limitations under the License.
...
+				}
+				fmt.Print(string(resp))
+				return nil
+			}
+			return nil
 		},
 	}
 
+	// define flags
+	cmd.Flags().StringP("host", "h", "", "host URL")
```
Comment (0xff-dev) [inline pkg/depository/create.go:145]: Do not use the `h` shorthand, which conflicts with the help parameter.

File: pkg/depository/create.go (hunk 11 above)
Comment (0xff-dev) [reply pkg/depository/create.go:132]: The `postValue` should be placed in the body, with the request header set.
In the above code, the request header `Content-Type:application/x-www-form-urlencoded` should be set

```go
postValue := url.Values{}
postValue.Add("message", msgBase64)
postValue.Add("value", valueBase64)

// -> http://localhost/basic/xxxx
host = fmt.Sprintf("%s%s", host, common.CreateDepository)
body := bytes.NewBuffer([]byte(postValue.Encode()))
uhttp.Do(host, http.MethodPost, map[string]string{
"Content-Type": "application/x-www-form-urlencoded",
}, body)
```

File: pkg/depository/create.go [hunk 13]
```diff
@@ -17,18 +17,243 @@ limitations under the License.
...
+					fmt.Print("putting untrusted value...\n")
+					postValue := url.Values{}
+					postValue.Add("value", valueBase64)
+
+					// -> http://localhost/basic/putUntrustValue
+					host = fmt.Sprintf("%s%s", host, common.CreateUntrustedDepository)
+					body := bytes.NewBuffer([]byte(postValue.Encode()))
+					resp, err := uhttp.Do(host, http.MethodPost, map[string]string{
+						"Content-Type": "application/x-www-form-urlencoded",
+					}, body.Bytes())
```
Comment (0xff-dev) [inline pkg/depository/create.go:128]: If the function argument is a `[]byte`, use `[]byte(postValue.Encode())` directly.

File: pkg/depository/create.go [hunk 14]
```diff
@@ -17,18 +17,241 @@ limitations under the License.
...
+			}
+
+			valueBase64 := generateValueDepotBase64(n, t, id, p)
+
+			if !accGen {
+				// FIXME: key auth to be done, bypassed temporally
+				// key auth
+				var privKey ecdsa.PrivateKey
+				err = json.Unmarshal([]byte(key), &privKey)
+				if err != nil {
```
Comment (0xff-dev) [inline pkg/depository/create.go:91]: 1. Is the `key` a filename or a private key string?
2. Is it correct to use json to parse the ecdsa private key?

```go
    a, _ := pem.Decode([]byte(x))
    pk, err := x509.ParseECPrivateKey(a.Bytes)
    if  err != nil {
        log.Fatal("ParseECPrivateKey error ", err)
    }
```

Comment (Lanture1064) [reply pkg/depository/create.go:91]: I assumed `key` is a password-like `string` before so I handled it with `json`. Should the `privateKey` be saved into & parsed from local file?

Comment (bjwswang) [reply pkg/depository/create.go:91]: @Lanture1064 @0xff-dev as we discussed, we manage key paris with another command `bc-cli  create/get/delete account ` in https://github.com/bestchains/bc-cli/issues/8

**Review by bjwswang — CHANGES_REQUESTED** [review]
place these changes into one commit

File: pkg/depository/create.go [hunk 15]
```diff
@@ -17,18 +17,188 @@ limitations under the License.
...
+		addr = library.ZeroAddress
+	}
+
+	return *privateKey, addr
+}
+
+func getNonce(account string) uint64 {
+	getReqValue := url.Values{}
+	getReqValue.Add("account", account)
+	host := fmt.Sprintf("%s?%s", common.GetDepository, getReqValue.Encode())
```
Comment (bjwswang) [inline pkg/depository/create.go:191]: this must be wrong. we use 
```shell
curl -X GET \
  http://localhost:9999/basic/currentNonce?account=xxx 
```

to get account's current nonce

**Review by bjwswang — CHANGES_REQUESTED** [review]

File: pkg/depository/create.go [hunk 16]
```diff
@@ -17,18 +17,188 @@ limitations under the License.
...
+				if err != nil {
+					return err
+				}
+				fmt.Print(string(resp))
+				return nil
+			} else {
+				// Generate Account
+				privKey, addr := randAccountAndPrivateKey()
+				// Get Nonce
+				non := getNonce(addr.String())
```
Comment (bjwswang) [inline pkg/depository/create.go:92]: `getNonce` is wrong

**Review by bjwswang — APPROVED** [review]

**Review by 0xff-dev — APPROVED** [review]

**Pull request merged**